File: R programs/MultiSpect_interval.R
MultiSpect_interval <- function(zt,output,alphalevel){
        attach(output)
        spectra <- list()
        coh <- list()
        nobs = dim(zt)[1]; dimen=dim(zt)[2]
        if(dimen==2){
                temp1 <- matrix(0,nfreq+1,nloop-nwarmup)
                temp2 <- matrix(0,nfreq+1,nloop-nwarmup)
                temp3 <- matrix(0,nfreq+1,nloop-nwarmup)
                spect11 <- array(0,c(nfreq_hat+1,nobs,3))
                spect22 <- array(0,c(nfreq_hat+1,nobs,3))
                spect22 <- array(0,c(nfreq_hat+1,nobs,3))
                coh21 <- array(0,c(nfreq_hat+1,nobs,3))
                for (j in 1:nobs){
                        if(j%%10==0){
                                cat("completed", j, "of", nobs, "observation", "\n")
                        }
                        for (p in 1:nloop){
                                if(p>nwarmup){
                                        if(length(xi[[nexp_curr[p]]][,p])==1){
                                                k=1
                                        }else{
                                                k = min(which(j <= xi[[nexp_curr[p]]][,p]))
                                        }
                                        temp1[,p-nwarmup] <- spect_hat[[nexp_curr[p]]][1,1,,k,p]
                                        temp2[,p-nwarmup] <- spect_hat[[nexp_curr[p]]][2,2,,k,p]
                                        temp3[,p-nwarmup] <- abs(spect_hat[[nexp_curr[p]]][2,1,,k,p])^2/
                                                (spect_hat[[nexp_curr[p]]][1,1,,k,p]*
                                                         spect_hat[[nexp_curr[p]]][2,2,,k,p])        
                                }
                        }
                        spect11[,j,1] <- apply(temp1,1,mean); spect11[,j,2] = apply(temp1,1,quantile, alphalevel/2); spect11[,j,3] = apply(temp1,1,quantile, 1-alphalevel/2)
                        spect22[,j,1] <- apply(temp2,1,mean); spect22[,j,2] = apply(temp2,1,quantile, alphalevel/2); spect22[,j,3] = apply(temp2,1,quantile, 1-alphalevel/2)
                        coh21[,j,1] <- apply(temp3,1,mean); coh21[,j,2] = apply(temp3,1,quantile, alphalevel/2); coh21[,j,3] = apply(temp3,1,quantile, 1-alphalevel/2)
                }
                return(list(spect11=spect11, spect22=spect22, coh21=coh21))
                
        }else if (dimen==3){
                temp1 <- matrix(0,nfreq+1,nloop-nwarmup)
                temp2 <- matrix(0,nfreq+1,nloop-nwarmup)
                temp3 <- matrix(0,nfreq+1,nloop-nwarmup)
                temp4 <- matrix(0,nfreq+1,nloop-nwarmup)
                temp5 <- matrix(0,nfreq+1,nloop-nwarmup)
                temp6 <- matrix(0,nfreq+1,nloop-nwarmup)
                spect11 <- array(0,c(nfreq_hat+1,nobs,3))
                spect22 <- array(0,c(nfreq_hat+1,nobs,3))
                spect33 <- array(0,c(nfreq_hat+1,nobs,3))
                coh21 <- array(0,c(nfreq_hat+1,nobs,3))
                coh31 <- array(0,c(nfreq_hat+1,nobs,3))
                coh32 <- array(0,c(nfreq_hat+1,nobs,3))
                for (j in 1:nobs){
                        if(j%%10==0){
                                cat("completed", j, "of", nobs, "observation", "\n")
                        }
                        for (p in 1:nloop){
                                if(p>nwarmup){
                                        if(length(xi[[nexp_curr[p]]][,p])==1){
                                                k=1
                                        }else{
                                                k = min(which(j <xi[[nexp_curr[p]]][,p]))
                                        }
                                        temp1[,p-nwarmup] <- spect_hat[[nexp_curr[p]]][1,1,,k,p]
                                        temp2[,p-nwarmup] <- spect_hat[[nexp_curr[p]]][2,2,,k,p]
                                        temp3[,p-nwarmup] <- spect_hat[[nexp_curr[p]]][3,3,,k,p]
                                        temp4[,p-nwarmup] <- abs(spect_hat[[nexp_curr[p]]][2,1,,k,p])^2/
                                                (spect_hat[[nexp_curr[p]]][1,1,,k,p]*
                                                         spect_hat[[nexp_curr[p]]][2,2,,k,p])
                                        temp5[,p-nwarmup] <- abs(spect_hat[[nexp_curr[p]]][3,1,,k,p])^2/
                                                (spect_hat[[nexp_curr[p]]][1,1,,k,p]*
                                                         spect_hat[[nexp_curr[p]]][3,3,,k,p])
                                        temp6[,p-nwarmup] <- abs(spect_hat[[nexp_curr[p]]][3,2,,k,p])^2/
                                                (spect_hat[[nexp_curr[p]]][3,3,,k,p]*
                                                         spect_hat[[nexp_curr[p]]][2,2,,k,p])
                                }

                        }
                        spect11[,j,1] <- apply(temp1,1,mean); spect11[,j,2] = apply(temp1,1,quantile, alphalevel/2); spect11[,j,3] = apply(temp1,1,quantile, 1-alphalevel/2)
                        spect22[,j,1] <- apply(temp2,1,mean); spect22[,j,2] = apply(temp2,1,quantile, alphalevel/2); spect22[,j,3] = apply(temp2,1,quantile, 1-alphalevel/2)
                        spect33[,j,1] <- apply(temp3,1,mean); spect33[,j,2] = apply(temp3,1,quantile, alphalevel/2); spect33[,j,3] = apply(temp3,1,quantile, 1-alphalevel/2)
                        coh21[,j,1] <- apply(temp4,1,mean); coh21[,j,2] = apply(temp4,1,quantile, alphalevel/2); coh21[,j,3] = apply(temp4,1,quantile, 1-alphalevel/2)
                        coh31[,j,1] <- apply(temp5,1,mean); coh31[,j,2] = apply(temp5,1,quantile, alphalevel/2); coh31[,j,3] = apply(temp5,1,quantile, 1-alphalevel/2)
                        coh32[,j,1] <- apply(temp6,1,mean); coh32[,j,2] = apply(temp6,1,quantile, alphalevel/2); coh32[,j,3] = apply(temp6,1,quantile, 1-alphalevel/2)
                }
                return(list(spect11=spect11, spect22=spect22, coh21=coh21))
        }
        detach(output)
}

File: matlab programs/MultiSpect_interval.m
function[spectra, coh] = MultiSpect_interval(zt, spect_matrices, fit, params, alphalevel) 

dim = size(zt); nobs = dim(1); dimen=dim(2);
spectra = cell(dimen,1);
coh = cell(dimen,1);
if dimen==3  
    
    temp1 = zeros(params.nfreq+1,(params.nloop-params.nwarmup));
    temp2 = zeros(params.nfreq+1,(params.nloop-params.nwarmup));
    temp3 = zeros(params.nfreq+1,(params.nloop-params.nwarmup));
    temp4 = zeros(params.nfreq+1,(params.nloop-params.nwarmup));
    temp5 = zeros(params.nfreq+1,(params.nloop-params.nwarmup));
    temp6 = zeros(params.nfreq+1,(params.nloop-params.nwarmup));
    spect11 = zeros(params.nfreq+1,nobs,3);
    spect22 = zeros(params.nfreq+1,nobs,3);
    spect33 = zeros(params.nfreq+1,nobs,3);
    coh21 = zeros(params.nfreq+1,nobs,3);
    coh31 = zeros(params.nfreq+1,nobs,3);
    coh32 = zeros(params.nfreq+1,nobs,3);
    for j=1:nobs
        if(mod(j,10)==0)
            fprintf('Completed: %g of %g observations \n' ,j, nobs)
        end
        for p=1:params.nloop
            if(p>params.nwarmup)
                if length(fit(fit(1).nexp_curr(p)).xi(:,p))==1
                    k=1;
                else    
                    k = min(find(j<=fit(fit(1).nexp_curr(p)).xi(:,p)));
                end    
                temp1(:,p-params.nwarmup) = squeeze(spect_matrices{fit(1).nexp_curr(p)}(1,1,:,k,p));
                temp2(:,p-params.nwarmup) = squeeze(spect_matrices{fit(1).nexp_curr(p)}(2,2,:,k,p));
                temp3(:,p-params.nwarmup) = squeeze(spect_matrices{fit(1).nexp_curr(p)}(3,3,:,k,p));
                temp4(:,p-params.nwarmup) = abs(squeeze(spect_matrices{fit(1).nexp_curr(p)}(2,1,:,k,p))).^2./...
                                (squeeze(spect_matrices{fit(1).nexp_curr(p)}(1,1,:,k,p)).*squeeze(spect_matrices{fit(1).nexp_curr(p)}(2,2,:,k,p)));
                temp5(:,p-params.nwarmup) = abs(squeeze(spect_matrices{fit(1).nexp_curr(p)}(3,1,:,k,p))).^2./...
                                (squeeze(spect_matrices{fit(1).nexp_curr(p)}(1,1,:,k,p)).*squeeze(spect_matrices{fit(1).nexp_curr(p)}(3,3,:,k,p)));        
                temp6(:,p-params.nwarmup) = abs(squeeze(spect_matrices{fit(1).nexp_curr(p)}(3,2,:,k,p))).^2./...
                                (squeeze(spect_matrices{fit(1).nexp_curr(p)}(2,2,:,k,p)).*squeeze(spect_matrices{fit(1).nexp_curr(p)}(3,3,:,k,p)));                         
            end
        end
        spect11(:,j,1) = mean(temp1,2); spect11(:,j,2:3) =  quantile(temp1,[alphalevel/2, 1-alphalevel/2],2);
        spect22(:,j,1) = mean(temp2,2); spect22(:,j,2:3) =  quantile(temp2,[alphalevel/2, 1-alphalevel/2],2);
        spect33(:,j,1) = mean(temp3,2); spect33(:,j,2:3) =  quantile(temp3,[alphalevel/2, 1-alphalevel/2],2);
        coh21(:,j,1) = mean(temp4,2); coh21(:,j,2:3) =  quantile(temp4,[alphalevel/2, 1-alphalevel/2],2);
        coh31(:,j,1) = mean(temp5,2); coh31(:,j,2:3) =  quantile(temp5,[alphalevel/2, 1-alphalevel/2],2);
        coh32(:,j,1) = mean(temp6,2); coh32(:,j,2:3) =  quantile(temp6,[alphalevel/2, 1-alphalevel/2],2);
    end    
    
    spectra{1}=spect11;spectra{2}=spect22;spectra{3}=spect33;
    coh{1}=coh21;coh{2}=coh31; coh{3}=coh32;
    
elseif dimen==2
  
    temp1 = zeros(params.nfreq+1,(params.nloop-params.nwarmup));
    temp2 = zeros(params.nfreq+1,(params.nloop-params.nwarmup));
    temp3 = zeros(params.nfreq+1,(params.nloop-params.nwarmup));
    spect11 = zeros(params.nfreq+1,nobs,3);
    spect22 = zeros(params.nfreq+1,nobs,3);
    coh21 = zeros(params.nfreq+1,nobs,3);
    for j=1:nobs
        if(mod(j,10)==0)
            fprintf('Completed: %g of %g observations \n' ,j, nobs)
        end
        for p=1:params.nloop
            if(p>params.nwarmup)
                if length(fit(fit(1).nexp_curr(p)).xi(:,p))==1
                    k=1;
                else    
                    k = min(find(j<=fit(fit(1).nexp_curr(p)).xi(:,p)));
                end    
                temp1(:,p-params.nwarmup) = squeeze(spect_matrices{fit(1).nexp_curr(p)}(1,1,:,k,p));
                temp2(:,p-params.nwarmup) = squeeze(spect_matrices{fit(1).nexp_curr(p)}(2,2,:,k,p));
                temp3(:,p-params.nwarmup) = abs(squeeze(spect_matrices{fit(1).nexp_curr(p)}(2,1,:,k,p))).^2./...
                                    (squeeze(spect_matrices{fit(1).nexp_curr(p)}(1,1,:,k,p)).*squeeze(spect_matrices{fit(1).nexp_curr(p)}(2,2,:,k,p))); 
            end
        end
        spect11(:,j,1) = mean(temp1,2); spect11(:,j,2:3) =  quantile(temp1,[alphalevel/2, 1-alphalevel/2],2);
        spect22(:,j,1) = mean(temp2,2); spect22(:,j,2:3) =  quantile(temp2,[alphalevel/2, 1-alphalevel/2],2);
        coh21(:,j,1) = mean(temp3,2); coh21(:,j,2:3) =  quantile(temp3,[alphalevel/2, 1-alphalevel/2],2);
    end    

    spectra{1}=spect11;spectra{2}=spect22;
    coh{1}=coh21;  

end









